File: n8n-plugin-worker/app/internal-auth.mjs
/** n8n 插件 Worker 的正文绑定 HMAC 请求验证。 */

import { createHash, createHmac, timingSafeEqual } from 'node:crypto'

const hex32 = /^[a-f0-9]{32}$/
const hex64 = /^[a-f0-9]{64}$/

export class InternalRequestVerifier {
  /** 保存独立密钥、时钟和进程内 nonce 缓存。 */
  constructor(secret, clock = () => Math.floor(Date.now() / 1000)) {
    this.secret = secret
    this.clock = clock
    this.used = new Map()
  }

  /** 校验目标、正文摘要和签名，并拒绝时间窗内 nonce 重放。 */
  verify(method, actualTarget, body, headers) {
    try {
      const timestampText = String(headers['x-internal-timestamp'] || '')
      const nonce = String(headers['x-internal-nonce'] || '')
      const target = String(headers['x-internal-target'] || '')
      const digest = String(headers['x-internal-content-sha256'] || '')
      const signature = String(headers['x-internal-signature'] || '')
      if (this.secret.length < 24 || !/^\d{1,12}$/.test(timestampText) || !hex32.test(nonce)
        || !hex64.test(digest) || !hex64.test(signature) || target !== actualTarget) return false
      const now = this.clock(); const signedAt = Number(timestampText)
      if (!Number.isSafeInteger(signedAt) || Math.abs(now - signedAt) > 60) return false
      const actualDigest = createHash('sha256').update(body).digest('hex')
      if (!equal(digest, actualDigest)) return false
      const canonical = `${String(method).toUpperCase()}\n${target}\n${timestampText}\n${nonce}\n${digest}`
      const expected = createHmac('sha256', this.secret).update(canonical).digest('hex')
      if (!equal(signature, expected)) return false
      for (const [value, usedAt] of this.used) if (usedAt < now - 60) this.used.delete(value)
      if (this.used.has(nonce)) return false
      this.used.set(nonce, now)
      return true
    } catch { return false }
  }
}

/** 对固定长度十六进制摘要执行常量时间比较。 */
function equal(left, right) {
  const first = Buffer.from(left, 'ascii'); const second = Buffer.from(right, 'ascii')
  return first.length === second.length && timingSafeEqual(first, second)
}

File: dify-plugin-worker/app/internal_auth.py
"""Dify 插件 Worker 的内部 HMAC 请求验证。"""

import hashlib
import hmac
import re
import threading
import time
from collections.abc import Mapping


HEX_32 = re.compile(r"^[a-f0-9]{32}$")
HEX_64 = re.compile(r"^[a-f0-9]{64}$")


class InternalRequestVerifier:
    """校验正文绑定签名、短时钟窗和 nonce 防重放。"""

    def __init__(self, secret: str) -> None:
        """创建每个 Worker 进程独立的 nonce 缓存。"""
        self.secret = secret
        self.used: dict[str, int] = {}
        self.lock = threading.Lock()

    def verify(self, method: str, actual_target: str, body: bytes,
               headers: Mapping[str, str], now: int | None = None) -> bool:
        """验证 HMAC 并在签名有效后原子登记 nonce。"""
        try:
            timestamp_text = headers.get("X-Internal-Timestamp", "")
            nonce = headers.get("X-Internal-Nonce", "")
            target = headers.get("X-Internal-Target", "")
            digest = headers.get("X-Internal-Content-SHA256", "")
            signature = headers.get("X-Internal-Signature", "")
            if (len(self.secret) < 24 or not timestamp_text.isdigit() or len(timestamp_text) > 12
                    or not HEX_32.fullmatch(nonce) or not HEX_64.fullmatch(digest)
                    or not HEX_64.fullmatch(signature) or target != actual_target):
                return False
            seconds = int(time.time()) if now is None else now
            signed_at = int(timestamp_text)
            if abs(seconds - signed_at) > 60 or not hmac.compare_digest(hashlib.sha256(body).hexdigest(), digest):
                return False
            canonical = f"{method.upper()}\n{target}\n{signed_at}\n{nonce}\n{digest}".encode()
            expected = hmac.new(self.secret.encode(), canonical, hashlib.sha256).hexdigest()
            if not hmac.compare_digest(expected, signature):
                return False
            with self.lock:
                self.used = {key: value for key, value in self.used.items() if value >= seconds - 60}
                if nonce in self.used:
                    return False
                self.used[nonce] = seconds
            return True
        except (TypeError, ValueError):
            return False
